scalyr_agent/copying_manager/copying_manager.py

```diff
@@ -72,6 +73,64 @@
 WORKER_SESSION_PROCESS_MONITOR_ID_PREFIX = "agent_worker_session_"
 
 
+class DynamicWorkers(object):
+    """
+    This class is responsible for managing the dynamic workers.
+    """
+
+    def __init__(self):
+        self.__workers_by_api_key = {}
+        self.__workers_by_id = {}
+        self.__last_id = 0
+        self.__worker_lock = threading.Lock()
+
+    def get_worker(self, worker_id):
+        # type: (str) -> Optional[CopyingManagerWorker]
+        """
+        Get worker by id.
+        """
+        return self.__workers_by_id[worker_id]
+
+    def get_or_create_worker(self, api_key, global_config):
+        # type: (str, Configuration) -> CopyingManagerWorker
+        """
+        Add worker.
+        """
+        if not api_key in self.__workers_by_api_key:
+            with self.__worker_lock:
+                if not api_key in self.__workers_by_api_key:
```

Comment:
If conditional seems to be duplicated (outside and inside the lock)? Can we just remove the outside conditional and only have one (with the lock)?

Reply:
We need both to avoid race condition in case the worker creation started between lines 99 and 100 in another thread.